File: winui/CellposeGUI.Tests/SegNpyIOTests.cs
using CellposeGUI.Services;
using Xunit;

namespace CellposeGUI.Tests;

public sealed class SegNpyIOTests
{
    private static string FixturePath(string name)
    {
        var projectDir = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "..", "..", ".."));
        return Path.GetFullPath(Path.Combine(projectDir, "..", "..", "tests", "fixtures", "seg_npy", name));
    }

    public static IEnumerable<object[]> FixtureCases =>
    [
        ["minimal_seg.npy", 2, "image.tif"],
        ["legacy_gui_seg.npy", 2, "legacy_image.tif"],
        ["with_instance_classes_seg.npy", 2, "image.tif"],
    ];

    [Theory]
    [MemberData(nameof(FixtureCases))]
    public void ReadCommittedFixtures(string fixtureName, int expectedNcells, string expectedSourceSuffix)
    {
        var path = FixturePath(fixtureName);
        Assert.True(File.Exists(path), $"missing fixture: {path}");

        var payload = SegNpyIO.Read(path);
        var filename = payload.GetValueOrDefault("filename")?.ToString() ?? "";
        Assert.EndsWith(expectedSourceSuffix, filename);

        Assert.IsType<SegNpyArray>(payload.GetValueOrDefault("masks"));
        var masks = (SegNpyArray)payload["masks"]!;
        Assert.Equal(expectedNcells, MaxLabel(masks));
    }

    [Fact]
    public void MinimalFixtureHasFlows()
    {
        var payload = SegNpyIO.Read(FixturePath("minimal_seg.npy"));
        Assert.IsType<List<object?>>(payload.GetValueOrDefault("flows"));
        var flows = (List<object?>)payload["flows"]!;
        Assert.Equal(2, flows.Count);
        Assert.All(flows, item => Assert.IsType<SegNpyArray>(item));
    }

    [Fact]
    public void InstanceClassesFixture()
    {
        var payload = SegNpyIO.Read(FixturePath("with_instance_classes_seg.npy"));
        Assert.IsType<SegNpyArray>(payload.GetValueOrDefault("instance_classes"));
        var classes = SegNpyIO.ReadLabels((SegNpyArray)payload["instance_classes"]!);
        Assert.Equal([1, 2], classes);
    }

    [Fact]
    public void LegacyFixtureHasColors()
    {
        var payload = SegNpyIO.Read(FixturePath("legacy_gui_seg.npy"));
        Assert.IsType<SegNpyArray>(payload.GetValueOrDefault("colors"));
        var colors = SegNpyIO.ReadColors((SegNpyArray)payload["colors"]!);
        Assert.Equal(2, colors.Length);
        Assert.Equal([100, 150, 200], colors[0]);
    }

    private static int MaxLabel(SegNpyArray masks)
    {
        var labels = SegNpyIO.ReadLabels(masks);
        return labels.Length == 0 ? 0 : labels.Max();
    }
}

File: swiftui/CellposeGUICore/Tests/CellposeGUICoreTests/SegNpyIOTests.swift
import CellposeGUICore
import XCTest

final class SegNpyIOTests: XCTestCase {
    private func fixturePath(_ name: String) -> String {
        var url = URL(fileURLWithPath: #filePath).deletingLastPathComponent()
        while url.path != "/" {
            let candidate = url.appendingPathComponent("tests/fixtures/seg_npy/\(name)")
            if FileManager.default.fileExists(atPath: candidate.path) {
                return candidate.path
            }
            url.deleteLastPathComponent()
        }
        fatalError("missing fixture: \(name)")
    }

    func testMinimalFixture() throws {
        let payload = try SegNpyIO.read(path: fixturePath("minimal_seg.npy"))
        guard let masks = payload["masks"] as? SegNpyArray else {
            XCTFail("missing masks")
            return
        }
        XCTAssertEqual(2, try SegNpyIO.readLabels(masks).max())
    }

    func testLegacyFixture() throws {
        let payload = try SegNpyIO.read(path: fixturePath("legacy_gui_seg.npy"))
        guard let masks = payload["masks"] as? SegNpyArray else {
            XCTFail("missing masks")
            return
        }
        XCTAssertEqual(2, try SegNpyIO.readLabels(masks).max())
    }

    func testInstanceClassesFixture() throws {
        let payload = try SegNpyIO.read(path: fixturePath("with_instance_classes_seg.npy"))
        guard let classesArray = payload["instance_classes"] as? SegNpyArray else {
            XCTFail("missing instance_classes")
            return
        }
        XCTAssertEqual([1, 2], try SegNpyIO.readLabels(classesArray))
    }
}
